### ISSUE TITLE
fluent react error with email input wrapped in <Localized>in MobileDialog

### ISSUE THREAD

**Issue by meandavejustice** [mozilla/testpilot#3631]

**Comment by devinreams** [comment]
To consider: fluent had a major release since the 0.4 series so if you go down this path it may require going through the entire upgrade path to get up to (at least) 0.6.

**Comment by ghost** [comment]
Dave is going to ping stas for help :)

**Comment by clouserw** [comment]
Marking this as size:s to try upgrading to 0.6.  If that doesn't work, we might need to change the size.

**Comment by lmorchard** [comment]
Is this the error? (Since I just saw it on a fresh new clone of the site)
```
Error: input is a void element tag and must neither have `children` nor use `dangerouslySetInnerHTML`.
    in input (created by NewsletterForm)
    in Localized (created by NewsletterForm)
    in form (created by NewsletterForm)
    in NewsletterForm (created by NewsletterFooter)
    in div (created by LayoutWrapper)
    in LayoutWrapper (created by NewsletterFooter)
    in div (created by LayoutWrapper)
    in LayoutWrapper (created by NewsletterFooter)
    in div (created by NewsletterFooter)
    in NewsletterFooter (created by View)
    in section (created by View)
    in View (created by HomePageNoAddon)
    in section (created by HomePageNoAddon)
    in HomePageNoAddon (created by HomePage)
    in HomePage (created by Route)
    in LocalizationProvider (created by App)
    in div (created by App)
    in App (created by Connect(App))
    in Connect(App) (created by Route)
    in Route
    in Switch
    in Router (created by BrowserRouter)
    in BrowserRouter
    in Provider
```

**Comment by meandavejustice** [comment]
Tried Version `0.7.0` first, which creates this error when trying to run our build

```
> testpilot@1.0.0 build:gulp /mnt/c/Users/Dave Justice/Code/testpilot
> gulp build

[14:05:35] Failed to load external module @babel/register
[14:05:35] Requiring external module babel-register
/mnt/c/Users/Dave Justice/Code/testpilot/node_modules/fluent-react/compat.js:452
  var TEMPLATE = document.createElement("template");
                 ^

ReferenceError: document is not defined
    at /mnt/c/Users/Dave Justice/Code/testpilot/node_modules/fluent-react/compat.js:452:18
    at /mnt/c/Users/Dave Justice/Code/testpilot/node_modules/fluent-react/compat.js:3:66
    at Object.<anonymous> (/mnt/c/Users/Dave Justice/Code/testpilot/node_modules/fluent-react/compat.js:6:2)
    at Module._compile (module.js:652:30)
    at Module._extensions..js (module.js:663:10)
    at Object.require.extensions.(anonymous function) [as .js] (/mnt/c/Users/Dave Justice/Code/testpilot/node_modules/babel-register/lib/node.js:152:7)
    at Module.load (module.js:565:32)
    at tryModuleLoad (module.js:505:12)
    at Function.Module._load (module.js:497:3)
    at Module.require (module.js:596:17)
```

Error with `0.6.0` is the same, just different line numbers in the stack trace.

```
> testpilot@1.0.0 build:gulp /mnt/c/Users/Dave Justice/Code/testpilot
> gulp build

[14:14:51] Failed to load external module @babel/register
[14:14:51] Requiring external module babel-register
/mnt/c/Users/Dave Justice/Code/testpilot/node_modules/fluent-react/compat.js:411
var TEMPLATE = document.createElement('template');
               ^

ReferenceError: document is not defined
    at /mnt/c/Users/Dave Justice/Code/testpilot/node_modules/fluent-react/compat.js:411:16
    at /mnt/c/Users/Dave Justice/Code/testpilot/node_modules/fluent-react/compat.js:3:65
    at Object.<anonymous> (/mnt/c/Users/Dave Justice/Code/testpilot/node_modules/fluent-react/compat.js:6:2)
    at Module._compile (module.js:652:30)
    at Module._extensions..js (module.js:663:10)
    at Object.require.extensions.(anonymous function) [as .js] (/mnt/c/Users/Dave Justice/Code/testpilot/node_modules/babel-register/lib/node.js:152:7)
    at Module.load (module.js:565:32)
    at tryModuleLoad (module.js:505:12)
    at Function.Module._load (module.js:497:3)
    at Module.require (module.js:596:17)
```

Bumping the size label for this up to medium now.

**Comment by lmorchard** [comment]
I think what we've hit is this upstream issue in fluent-react:

The use of `<template>` prevents server-side rendering
https://github.com/projectfluent/fluent.js/issues/175

**Comment by stasm** [comment]
Sorry that you're having trouble with Fluent. I'll look into the original bug and try to see if it's something I can fix on the `fluent-react` side. Is the error the one mentioned in https://github.com/mozilla/testpilot/issues/3631#issuecomment-411102984?

There are two packages which are related: `fluent` which formats translations defined in FTL files, and `fluent-react` which provides bindings for React.

`fluent-react` is at 0.7.0 right now. We have a fix for the `document is not defined` issue which I need to review and polish. Expect a new release soon.

`fluent` has just recently had its 0.7.0 version published, too. Test Pilot still uses `fluent` 0.4.3. I'll file a new issue to track updating it.

**Comment by stasm** [comment]
I'm trying to set up Test Pilot locally to reproduce the error, but I'm hitting some `optipng`-related build errors.

There's an example in `fluent-react` of how `<Localized>` can be used to localize text inputs with placeholders:

https://github.com/projectfluent/fluent.js/blob/8ac774dd8ef1ab46b6bce8083b891898b25846f8/fluent-react/examples/text-input/src/App.js#L35-L42

I just checked that it works correctly with `fluent-react` 0.7.0.  I'll report back after testing 0.4.1 which Test Pilot uses.

**Comment by lmorchard** [comment]
We haven't been able to upgrade to fluent-react 0.7.0 so far due to our server-side rendering running into https://github.com/projectfluent/fluent.js/issues/175

So, what I'm trying to do in PR #3783 is to at least work around our error, until / unless we can upgrade (and assuming the upgrade fixes the issue)

**Comment by meandavejustice** [comment]
Thanks @stasm, yes that is the correct error.


We are actually doing the same thing as the example you've linked.

https://github.com/mozilla/testpilot/blob/master/frontend/src/app/components/NewsletterForm/index.js#L49

``` js
<Localized id="newsletterFormEmailPlaceholder" attrs={{placeholder: true}}>
        <input
          type='email'
          placeholder="Your email here"
          required
          value={this.props.email}
          onChange={this.handleEmailChange}
        />
</Localized>
```

Feel free to ping me on slack or irc "JSON_voorhees" if you want help getting the local setup going

**Comment by meandavejustice** [comment]
@lmorchard stasm said above that they have a fix for the undefined document issue

>fluent-react is at 0.7.0 right now. We have a fix for the document is not defined issue which I need to review and polish. Expect a new release soon.

https://github.com/mozilla/testpilot/issues/3631#issuecomment-411419408

**Comment by stasm** [comment]
I was able to set up Test Pilot locally this morning. It wasn't too hard, actually. The `optipng` issues were related to npm running on wrong permissions.

I can browse to the main page in English and in French, and I don't see any errors in the console. The text input renders correctly. Would you mind providing steps to reproduce this bug?

**Comment by stasm** [comment]
The error mentioned in a comment above usually happens when a buggy translation is trying to insert some text content into an `<input>` element wrapped in a `<Localized>`. This could happen if the localization mistakenly sets a value on the message used to localize the `<input>`.

```properties
# ERROR: This message has both a value (Hello) and an attribute.
# The value ends up as a text child of the element wrapped in <Localized>.
# React doesn't allow `<input>` to have text content.
newsletterFormEmailPlaceholder = Hello
    .placeholder = Your email here
```

I verified all files in `locales/` and I don't see any such error. This is a long shot, but perhaps you're seeing this due to a locally-modified `app.ftl` in your working copy?

`fluent-react` 0.7.0 [has a fix](https://github.com/projectfluent/fluent.js/blob/master/fluent-react/CHANGELOG.md#fluent-react-070-may-18-2018) which protects void elements like `<input>` from being broken by a buggy translation. I'll try to unblock the server-side rendering issue this week. In the meantime, the work-around in #3783 should work fine, provided this isn't caused by a bug in the local setup :)

**Comment by stasm** [comment]
Oh wait, I was looking at `newsletterFormEmailPlaceholder ` because the work-around in #3783 touches it, but now I realize that the summary of this issue mentions `MobileDialog`. I found the following code with `<Localized>` wrappers commented out:

https://github.com/mozilla/testpilot/blob/5789c3c8db56425cc0cb6c683c5d7578b5c53848/frontend/src/app/components/MobileDialog/index.js#L205-L223

Are they disabled because they were causing this bug? If so, that's because their translations don't have attributes, and instead try to insert the translation as a text child into the `<input>` elements.

https://github.com/mozilla/testpilot/blob/7c552ad7b6ecd53c176cb18ca9d1b6b0f0f31e1f/locales/en-US/app.ftl#L91-L92

This should read:

```properties
mobileDialogPlaceholder =
    .placeholder = Enter your Email
mobileDialogPlaceholderSMS =
    .placeholder = Enter your Phone/Email
```

**Comment by meandavejustice** [comment]
picking this up next week after #3818 is merged
